### PR TITLE
[enhancement](cloud) optimize block cache lock

### PR DESCRIPTION
**Pull request by freemandealer**
1. async deletion when do stale rowsets reclycle
2. minimize lock critical size
3. add cache lock held & wait time info for debug

### CODE REVIEW

**Comment by freemandealer** [comment]
run buildall

**Comment by freemandealer** [comment]
run buildall

**Comment by freemandealer** [comment]
run buildall

**Comment by freemandealer** [comment]
run buildall

File: be/src/io/cache/block_file_cache.h [hunk 1]
```diff
@@ -27,15 +28,51 @@
...
+    std::chrono::time_point<std::chrono::steady_clock> start_time =                               \
+            std::chrono::steady_clock::now();                                                     \
+    std::lock_guard cache_lock(MUTEX);                                                            \
+    std::chrono::time_point<std::chrono::steady_clock> acq_time =                                 \
+            std::chrono::steady_clock::now();                                                     \
+    auto duration =                                                                               \
+            std::chrono::duration_cast<std::chrono::milliseconds>(acq_time - start_time).count(); \
+    if (duration > config::cache_lock_long_tail_threshold)                                        \
+        LOG(WARNING) << "Lock wait time " << std::to_string(duration) << "ms. "                   \
+                     << get_stack_trace_by_boost() << std::endl;                                  \
```
Comment (dataroaring) [inline be/src/io/cache/block_file_cache.h:47]: We should limit frequency of the log.

**Comment by freemandealer** [comment]
run buildall

File: be/src/io/cache/block_file_cache.h [hunk 2]
```diff
@@ -446,6 +489,9 @@ class BlockFileCache {
     LRUQueue _disposable_queue;
     LRUQueue _ttl_queue;
 
+    // keys for async remove
+    std::shared_ptr<boost::lockfree::spsc_queue<FileCacheKey>> _recycle_keys;
```
Comment (gavinchou) [inline be/src/io/cache/block_file_cache.h:493]: it seems there is no contention on _recycle_keys.
it is better to  use regular data structure instead for better understanding and maintainance

File: be/src/common/config.h [hunk 3]
```diff
@@ -984,6 +984,9 @@ DECLARE_mInt64(nodechannel_pending_queue_max_bytes);
 
 // The batch size for sending data by brpc streaming client
 DECLARE_mInt64(brpc_streaming_client_batch_bytes);
+DECLARE_mInt64(block_cache_wait_timeout_ms);
+DECLARE_mInt64(cache_lock_long_tail_threshold);
+DECLARE_Int64(file_cache_recycle_keys_size);
```
Comment (gavinchou) [inline be/src/common/config.h:989]: does it mean capacity in number of elements?

File: be/src/io/cache/block_file_cache.cpp [hunk 4]
```diff
@@ -1302,8 +1347,18 @@ void BlockFileCache::remove(FileBlockSPtr file_block, T& cache_lock, U& block_lo
     *_num_removed_blocks << 1;
 }
 
+void BlockFileCache::recycle_stale_rowset_async_bottom_half() {
+    FileCacheKey key;
+    while (_recycle_keys->pop(key)) {
+        Status st = _storage->remove(key);
+        if (!st.ok()) {
+            LOG_WARNING("").error(st);
```
Comment (gavinchou) [inline be/src/io/cache/block_file_cache.cpp:1355]: put some key word in the log e.g. 
"failed to remove when recycle_stale_rowset_async_bottom_half, key = {}"  ...

File: be/src/io/cache/block_file_cache.cpp [hunk 5]
```diff
@@ -1285,9 +1315,24 @@ void BlockFileCache::remove(FileBlockSPtr file_block, T& cache_lock, U& block_lo
...
+            // the file will be deleted in the bottom half
+            // so there will be a window that the file is not in the cache but still in the storage
+            // but it's ok, because the rowset is stale already
+            // in case something unexpected happen, set the _recycle_keys queue to zero to fallback
+            bool ret = _recycle_keys->push(key);
+            if (!ret) {
+                LOG_WARNING("Failed to push recycle key to queue, do it synchronously");
+                Status st = _storage->remove(key);
+                if (!st.ok()) {
+                    LOG_WARNING("").error(st);
```
Comment (gavinchou) [inline be/src/io/cache/block_file_cache.cpp:1333]: put some key word in the log e.g.
"failed to remove when recycle_stale_rowset_async_bottom_half, key = {}" ...

**Review by gavinchou — APPROVED** [review]

**Review by Hastyshell — APPROVED** [review]
LGTM

**Pull request merged**